README.md
# UwULang

[![Banner](./img/banner.png)](https://uwulang.vercel.app/)

[![Sanity Check](https://github.com/UwULang/uwulang/actions/workflows/sanity_check.yaml/badge.svg)](https://github.com/UwULang/uwulang/actions/workflows/sanity_check.yaml) [![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT) ![GitHub Repo stars](https://img.shields.io/github/stars/UwULang/uwulang?style=social)

The soon to be #1 programming language in the world. It is a turing-complete language modelled after the measured UwUness of a programming language called BrainFuck. Check out the [website](https://uwulang.vercel.app/) for more info! Alternatively, check out the [documentation](https://uwulang-docs.vercel.app/) for more info.

## Try it Out

### Install

If you are on Debian or Ubuntu, you can install the package from the [release](https://github.com/UwULang/uwulang/releases) and do `sudo apt install ./uwulang_<VERSION_NUMBER>_all.deb` replacing `<VERSION_NUMBER>` with the version number of the release.

Otherwise you can build it from source. First clone the repo:

```bash
git clone https://github.com/UwULang/uwulang.git
```

Then run

```bash
make install
```

to install the package to your local machine. This will install the `uwulang` binary to your local machine and a man page for the `uwulang` command. (requires make and gcc)

### Only Build Locally

Running it by first building it in c with `make` or just use `gcc src/main.c -Wall -std=c99 -o uwulang`. Once you have the binary, if you provide a file as the first arg, it will take the contents of the file and run UwULang on it:

```bash
./uwulang test/hellOwOrld.uwu
```

Otherwise if you provide no args, it will launch into the UwULang shell interpreter.

```bash
./uwulang
```

## Usage

The interpreter is a command-line program that reads a UwULang script from a file or from standard input and executes it. The interpreter should be able to be invoked with the command `uwulang`. with no arguments provided

That will launch you into an interactive interpreter. The interpreter should print a prompt starting with `> `, and then read a line of input from standard input. The interpreter should then execute the line of input as a UwULang program. The interpreter should then print the prompt again and read another line of input. This process should continue until the user does Ctrl-C or Ctrl-Z. The interpreter should then exit.

To run a UwULang script from a file, you can run `uwulang <filename>`. The interpreter should then execute the script in the file and exit.

By using the `-v` flag, you can get the version of UwULang
Using the `-h` flag, you can get the help menu

There is also a helpful man page! You can access it by running `man uwulang`

## Instructions

- 👆 - increment
- 👇 - decrement
- 👉- go right
- 👈- go left
- 🥺- print char
- 😳- get char
- 🥴- random short
- 😒- jump to 😡 if curr == 0
- 😡- jump back to 😒 if curr != 0

\*definitely not just [brainfuck](https://esolangs.org/wiki/Brainfuck) with extended functionality that you can use [uwufier](https://github.com/Zeyu-Li/uwufier) conversion tool

## Specification file

If you want to implement UwULang, you can use the specification file `uwulang.md`

### Hello World

Print Hello World to the console

```uwu
👆👆👆👆👆👆👆👆😒👉👆👆👆👆😒👉👆👆👉👆👆👆👉👆👆👆👉👆👈👈👈👈👇😡👉👆👉👆👉👇👉👉👆😒👈😡👈👇😡👉👉🥺👉👇👇👇🥺👆👆👆👆👆👆👆🥺🥺👆👆👆🥺👉👉🥺👈👇🥺👈🥺👆👆👆🥺👇👇👇👇👇👇🥺👇👇👇👇👇👇👇👇🥺👉👉👆🥺👉👆👆🥺
```

### First 10000 square numbers

Print the first 10000 square numbers

```uwu
👆👆👆👆😒👉👆👆👆👆👆👈👇😡👉😒👈👆👆👆👆👆👉👇😡👆👈👆😒👉😒👉👆👉👆👈👈👇😡👆👆👉👉😒👈👈👆👉👉👇😡👉👉👉😒👇😡👆👆👉😒👇😡👆👉👉👉👆😒😒👇😡👆👆👆👆👆👆👉👉👉😡👈👈👈😒😒👈👆👆👆👆👆👆👆👆👈👆👆👉👉👇😡👆👈🥺👈😒👉👇👇👇👇👈👇😡👈😡👈👈😒👉👉👉👉👉😒👉👉👉😒👇😡👆👆👆👆👆👆👆👆👆👈😒👉👇👈👇😡👆👆👆👆👆👆👆👆👆👉😒👇😒👈👇👉👇😡👆😒👈👈👈😡😡👈😒👉👆👈👇😡👉😡👈👈👇😡👈👈👇😡
```

### E (transcendental number)

Prints digits of the transcendental number e (this program is non terminating ie it will run forever for ever greater precision)

```uwu
👉👉👉👉👆👆👉👆👉👆👆👉👆👉👉👆👆👈👆😒😒👉😒👉👉😒👉👉👉👉😡👈👈👈👈😒😒👉👉👉👉👆👈👈👈👈👇😡👈👈👈👈😡👉👉👉👉👉👉😡👆👈😡👉👇👉👉👇👇😒👆😒👆👆👆👈👈👈👈👇👇😡👆👆👉👉👉👉👇👇😡👆😒👉👉👉👉😡👈👈👈👈😒👈👈👆👈👆👈😡👈👈😒👉👉👉👉👉👉😒😒👈👈👈👈👆👉👉👉👉👇😡👉👉👉👉😡👈👈👈👈👈👈👈👈😒👈👈👈👈😡👉👉👇😒👈👈👆👉👉👇😡👆👈👈😒👇👉👉👉👉😒👇😒👆👉👉👉👉👇😡👇👈👈👇😒👉👉👉👉👇😡👆👆👉👉👆😒👇👈👈👈👈👆😡👆👉👉👉👉😡👈👈👈👈😒👈👈👈👈😡😡👉😒👇😒👈👆👉👇😡😡👆👈😒👇👉👉👉👉😒👇😒👆👉👉👉👉👇😡👇👈👈👈👇😒👉👉👉👉👇😡👆👆👉👉👉👆😒👇👈👈👈👈👆😡👆👉👉👉👉😡👈👈👈👈😒👈👈👈👈😡😡👈👈😡👉👉👉👆😒👉👉👉👉😡👇😒👆👈👈👈👈👇👇😡👆👆😒👈👈👈👈😡👉👉👉👆😒👉👇😒👉👉😒👇👇😒👆👆👉👉👆👉👉👇👇😡👇👈😒👇😒👇😒👆👆👆👈👈👈👈👇😡👆👉👉👉👉👇😡😡👆👆👉👆😒👇👈👈👈👈👆😡👆👆👉👉👆👉👉😡👈👈😒👉😒👈👇👈👈👈😡👆👈😡👉👇👉👉👉😡👆👉😒👉👉👉👉😡👇😒👆👈👈👈👈👇👇😡👆👆👈😒😒👉👉👉👉😡👈👈👈👈😒👇😒👆👉😒👈👇👉👇😡👆👆👈😒😒👉👇👈👇😡👆👆😒👈👈👈👈😡👆👉👉👆👉👉👇😡👆👆👈👈👈👈👇😡👉👇😒👆😒👈👆😒👈👈👈👈😡👉😡👈👆👉😡👆👈😒👇👉👇👉👉👉😒👇😡😡👆👈👈👈👈😡😡👉😒👈👈👈👈😡👉😒👇😒👇😒👆👆👆👆👆😒👉👆👆👆👆👆👆👆👆👈👇😡👉👇🥺👉👉👉👇😒👈👈👈👇👇👇👇🥺👈😡👈😒👈👈😡👉👉😒👇😡👉👇👉👉👆😒😒👉👉👉👉😡👆😒👇😒👇👉👉👉👉👆👉👉👉👉👉👉👉👉👇😒👇😒👆👆👆👈👈👈👈😒👇😡😡👆👉👉👉👉👇😡👆👆😒👈👈👈👈😡😡👆👈👈👈👈😡👉👉👉😡👆👈👆👈👈😡👉😒👇😒👇👉😒👇👇😒👆👆👉👉👉👉👇👇😡👇👉😒👇😒👇😒👆👆👆👈👈👈👈👇😡👆👉👉👉👉👇😡😡👆👆👈👆😒👇👈👈👈👈👆😡👆👆👉👉👉👉😡👈👈👈👈😒👉😒👈👈👈👈😡👆👈😡👉👇👉👉😡👈😡👉👉👉👉😒👇👇😒👆👆👉👉👉👉👇👇😡👇👈👇👇😒👆👆👆👉👉👉👉👇👇😡👆👉👆😒👇👈👈👈👈👆😡👆👆👉👉👉👉😡👈👈👈👈👈😒👈👈👈👈😡👈😡👉😒👉👆👈👈👆👆👈😡👈😡👉😒👆👉😒👇👇😒👆👆👉👉👉👉👇👇😡👇👉👇👇😒👆👆👆👉👉👉👉👇👇😡👆👈👆😒👇👈👈👈👈👆😡👆👆👉👉👉👉😡👈👈👈😒👈👈👈👈😡😡👉👉😡👉😡
```

## Conversion

If you want to convert bf files to UwU files use the [Transpiler](https://uwulang.vercel.app/transpiler) or [UwUfier](https://github.com/UwULang/uwufier) written in Golang

## License

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v0.1.0

Initial release


test/old.c
/* 
* UwULang (MIT)
*/ 
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

#define STACK_SIZE      512

#define STACK_PUSH(A)   (STACK[SP++] = A)
#define STACK_POP()     (STACK[--SP])
#define STACK_EMPTY()   (SP == 0)
#define STACK_FULL()    (SP == STACK_SIZE)

static unsigned short STACK[STACK_SIZE];
static unsigned int SP = 0;

struct Block {
    char value;
    struct Block* next;
    struct Block* parent;
} UwU = {0, NULL, NULL};

struct Block* next(struct Block* curr) {
    if (curr->next != NULL) {
        return curr->next;
    } else {
        curr->next = malloc(sizeof (struct Block));
        curr->next->value = 0;
        // if next cannot be alloc
        if (curr->next == NULL)
            return NULL;
        else 
            return curr->next;
    }
}

/// Main takes in either a single file with file extension uwu
/// or if no args are present will run interperted on the command line
int main(int argc, char *argv[]) {
    struct Block* head = &UwU;
    struct Block* curr = &UwU;
    // Use current time as seed for random generator
    srand(time(0));

    // inital way
    unsigned char code[32768] = {0};

    unsigned char currInputChar;
    int initCounter = 0;
    int charCounter = 0;
    while ((currInputChar = getchar()) != EOF && currInputChar != '\n') {
        code[initCounter] = currInputChar;
        charCounter++;
        initCounter++;
    }
    // FILE *fptr;
    // // get from cin or file
    // if (argc == 1) {
    //     while ((currInputChar = getchar()) != EOF && currInputChar != '\n') {
    //         *inputCounter = currInputChar;
    //         inputCounter++;
    //         counter++;
    //     }
    // } else {
    //     // open file
    //     printf("%s", argv[1]);
    //     fptr = fprintf(argv[1],"%c", code);
    //     fclose(fptr);
    // }

    int jump_table[32768] = {0};
    int counter = 0;
    // hash map for pre-complied [] blocks
    for (int i = 0; i < charCounter / 4 + 1; i++) {
        if (code[counter] == 240 && code[counter+1] == 159 && code[counter+2] == 152 &&  code[counter+3] == 146) {
            STACK_PUSH(i);
        } else if (code[counter] == 240 && code[counter+1] == 159 && code[counter+2] == 152 &&  code[counter+3] == 161) {
            int start = STACK_POP();
            jump_table[start] = i;
            jump_table[i] = start;
        }
        // currChar == 152 && nextChar == 161
        counter+=4;
    }

    int ptr = 0;
    int codePtr = 0;
    int currChar, nextChar;
    while (code[ptr] != EOF && code[ptr] != '\n' && code[ptr] != '\0') {
        // printf("%d: %d %d %d %d\n", ptr, code[ptr], code[ptr + 1], code[ptr + 2], code[ptr + 3]);
        // if not emoji continue
        if (code[ptr] != 240) {
            ptr++;
            continue;
        }
        ptr++;
        if (code[ptr] != 159) {
            if (code[ptr] == '\n') break;
            ptr++;
            continue;
        }
        ptr++;
        // printf("%d", *pc);
        currChar = code[ptr];
        nextChar = code[ptr+1];
        // printf("%d", curr.value);
        if (currChar == 145 && nextChar == 134) {
            if (curr->value < 255) curr->value++;
            else curr->value = 0;
        }
        else if (currChar == 145 && nextChar == 135) {
            if (curr->value > 0) curr->value--;
            else curr->value = 255;
        }
        else if (currChar == 165 && nextChar == 186) {
            // putchar(curr->value);
            printf("%c", curr->value);
        }
        else if (currChar == 152 && nextChar == 179) curr->value = getchar();
        else if (currChar == 145 && nextChar == 137) curr = next(curr);
        else if (currChar == 145 && nextChar == 136) {
            // if parent exists return parent, otherwise return current
            curr = curr->parent ? curr->parent : curr;
        } else if (currChar == 165 && nextChar == 180) {
            // if parent exists return parent, otherwise return current
            curr->value = rand() % 128;
        } else if (currChar == 152 && nextChar == 146 && curr->value == 0) {
            // [
            printf("%d %d\n", jump_table[codePtr], codePtr);
            
            ptr = jump_table[codePtr] * 4 + 2; // get hashed value
            codePtr = jump_table[codePtr];
        } else if (currChar == 152 && nextChar == 161 && curr->value != 0) {
            // 😡 ]
            printf("%d %d\n", jump_table[codePtr], codePtr);
            ptr = jump_table[codePtr] * 4 + 2; // get hashed value
            codePtr = jump_table[codePtr];
        }

        ptr+=2;
        codePtr++;
    }

    // dealloc from head
    if (head->next != NULL) {
        struct Block* deletePtr = head->next;
        while ((deletePtr = deletePtr->next)) {
            free(deletePtr->parent);
        }
        free(deletePtr);
    }
    // since struct is statically defined, no need to dealloc

    // new line
    putchar('\n');


    return 0;
}

// tests
// >>++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++. 
// 👉👉👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆🥺


test/test.c
/* 
* UwULang (MIT)
*/ 
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

struct Block {
    char value;
    struct Block* next;
    struct Block* parent;
} UwU = {0, NULL, NULL};

struct Block* next(struct Block* curr) {
    if (curr->next != NULL) {
        return curr->next;
    } else {
        curr->next = malloc(sizeof (struct Block));
        curr->next->value = 0;
        // if next cannot be alloc
        if (curr->next == NULL)
            return NULL;
        else 
            return curr->next;
    }
}

/// Main takes in either a single file with file extension uwu
/// or if no args are present will run interperted on the command line
int main(int argc, char *argv[]) {
    struct Block* head = &UwU;
    struct Block* curr = &UwU;
    // Use current time as seed for random generator
    srand(time(0));

    // TODO: convert char to UTF-8 emojis

    // inital way
    char code[32768] = {0};
    int *inputCounter = code;
    int currInputChar;

    // testing emojis
redo:
    while ((currInputChar = getchar()) != EOF && currInputChar != '\n') {
        printf("%d ", currInputChar);
    }
    goto redo;

    // since struct is statically defined, no need to dealloc

    // new line
    putchar('\n');


    return 0;
}

// tests
// >>++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++. 👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆👆🥺


test/test.sh
#!/bin/bash
echo "Testing output:"
./uwulang test/hellOwOrld.uwu > test1.out
./uwulang test/sqwuare10000.uwu > test2.out

diff test1.out test/hello.out > diff.out
diff test2.out test/square.out >> diff.out

if [ -s diff.out ] # test if diff is empty
then
    cat diff.out # print
    exit 1 # test failed
fi

pseudocode/main.py
#!/usr/bin/env python3
import random
import sys

def main():
    code = ''
    if len(sys.argv) == 1: code = input()
    else:
        # try and open argv0
        try:
            with open(sys.argv[1], 'r', encoding="utf-8") as fp:
                code = fp.read()
        except:
            raise FileExistsError()

    # code
    OwO = [i for i in code]

    # blocks
    UwU = [0]
    ptr = 0

    # precompile
    front_ptr = 0
    back_ptr = len(OwO) - 1
    if code.count('😒') != code.count('😡'): raise Exception("Invalid")

    jump_table = dict()
    # help from https://github.com/pocmo/Python-Brainfuck/blob/master/brainfuck.py
    temp_bracestack = []
    for index, item in enumerate(OwO):
        if item == '😒':
            temp_bracestack.append(index)
        elif item == '😡':
            start = temp_bracestack.pop()
            # doubly indexed jump table
            jump_table[start] = index
            jump_table[index] = start

    code_ptr = 0
    while code_ptr < len(OwO):
        item = OwO[code_ptr]
        if item == '\n': return
        elif item == '👆': UwU[ptr] = UwU[ptr] + 1 if UwU[ptr] < 255 else 0
        elif item == '👇': UwU[ptr] = UwU[ptr] - 1 if UwU[ptr] > 0 else 255
        elif item == '😳': 
            userInput = input()[0]
            UwU[ptr] = ord(userInput)
        elif item == '🥺': print(chr(UwU[ptr]), end='')
        elif item == '🥴': UwU[ptr] = random.randint(0, 128)
        elif item == '👈':
            if ptr <= 0: 
                ptr = 0
            else: ptr-=1
        elif item == '👉':
            ptr += 1
            if ptr == len(UwU): UwU.append(0)
        # if block start and current is 0, go to end 😡
        elif item == '😒' and UwU[ptr] == 0: code_ptr = jump_table[code_ptr]
        # if block end and current is not 0, go to start
        elif item == '😡' and UwU[ptr] != 0: code_ptr = jump_table[code_ptr]

        code_ptr+=1

    return

if __name__ == "__main__":
    main()


src/main.c
/* 
* UwULang, an esoteric programming language with emojis
* More info on https://github.com/UwULang/uwulang 
* Author: @Zeyu-Li 
* License: MIT
*/ 
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

// program version
#define VERSION         "0.1.0"

#define STACK_SIZE      512

#define STACK_PUSH(A)   (STACK[SP++] = A)
#define STACK_POP()     (STACK[--SP])
#define STACK_EMPTY()   (SP == 0)
#define STACK_FULL()    (SP == STACK_SIZE)

static unsigned short STACK[STACK_SIZE];
static unsigned int SP = 0;

struct Block {
    unsigned char value;
    struct Block* next;
    struct Block* parent;
};

struct Block* next(struct Block* curr) {
    if (curr->next != NULL) {
        return curr->next;
    } else {
        curr->next = (struct Block*)malloc(sizeof (struct Block));
        // init next block
        curr->next->value = 0;
        curr->next->next = NULL;
        curr->next->parent = curr;
        // if next cannot be alloc
        if (curr->next == NULL) 
            return NULL;
        else 
            return curr->next;
    }
}

int event_loop(int argc, char *argv[]) {
    // Use current time as seed for random generator
    srand(time(0));
    FILE *fp;
    
    struct Block UwU = {0, NULL, NULL};

    struct Block* head = &UwU;
    struct Block* curr = &UwU;

    // inital way
    unsigned char code[32768] = {0};

    unsigned char currInputChar;
    int initCounter = 0;
    int charCounter = 0;
    
    // get from cin or file
    if (argc == 1) {
        printf("> ");
        while ((currInputChar = getchar()) != EOF && currInputChar != '\n') {
            code[initCounter] = currInputChar;
            charCounter++;
            initCounter++;
        }
    } else {
        // open file
        fp = fopen(argv[1], "r");
        if (!fp) {
            printf("Error in opening file!\n");
            return 1;
        }
        fscanf(fp, "%[^=]=%[^;]; ", code);
        charCounter = strlen((char*)code);
        fclose(fp);
    }
    

    int jump_table[32768] = {0};
    int counter = 0;
    // hash map for pre-complied [] blocks
    for (int i = 0; i < charCounter / 4 + 1; i++) {
        if (code[counter] == 240 && code[counter+1] == 159 && code[counter+2] == 152 && code[counter+3] == 146) {
            STACK_PUSH(i);
        } else if (code[counter] == 240 && code[counter+1] == 159 && code[counter+2] == 152 && code[counter+3] == 161) {
            int start = STACK_POP();
            jump_table[start] = i;
            jump_table[i] = start;
        }
        counter+=4;
    }

    int ptr = 0;
    int codePtr = 0;
    int currChar, nextChar;
    while (code[ptr] != EOF && code[ptr] != '\0') {
        // if not emoji continue
        if (code[ptr] != 240) {
            ptr++;
            continue;
        }
        ptr++;
        if (code[ptr] != 159) {
            ptr++;
            continue;
        }
        ptr++;

        currChar = code[ptr];
        nextChar = code[ptr+1];

        if (currChar == 145 && nextChar == 134) {
            if (curr->value < 255) curr->value++;
            else curr->value = 0;
        }
        else if (currChar == 145 && nextChar == 135) {
            if (curr->value > 0) curr->value--;
            else curr->value = 255;
        }
        else if (currChar == 165 && nextChar == 186) {
            // putchar(curr->value);
            printf("%c", curr->value);
        }
        else if (currChar == 152 && nextChar == 179) curr->value = getchar();
        else if (currChar == 145 && nextChar == 137) curr = next(curr);
        else if (currChar == 145 && nextChar == 136) {
            // if parent exists return parent, otherwise return current
            curr = curr->parent ? curr->parent : curr;
        } else if (currChar == 165 && nextChar == 180) {
            // if parent exists return parent, otherwise return current
            curr->value = rand() % 128;
        } else if (currChar == 152 && nextChar == 146 && curr->value == 0) {
            codePtr = jump_table[codePtr];
            ptr = codePtr * 4 + 2; // get hashed value
        } else if (currChar == 152 && nextChar == 161 && curr->value != 0) {
            // 😡 ]
            codePtr = jump_table[codePtr];
            ptr = codePtr * 4 + 2; // get hashed value
        }

        ptr+=2;
        codePtr++;
    }

    // dealloc from head
    if (head->next != NULL) {
        struct Block* deletePtr = head->next;
        while ((deletePtr = deletePtr->next)) {
            free(deletePtr->parent);
        }
        free(deletePtr);
    }
    // since struct is statically defined, no need to dealloc

    return 0;
}

/// Main takes in either a single file with file extension uwu
/// or if no args are present will run interpreted on the command line
int main(int argc, char *argv[]) {
    if (argc == 1) {
        // TODO: add support for multiple lines
        printf("Welcome to UwULang shell\nUwULang is the number 1 programming language\nFind out more at https://github.com/UwULang/uwulang\n");
        while (1) {
            event_loop(argc, argv);
        }
    } else {
        // TODO: add support for args

        // if -v or --version
        if (strcmp(argv[1], "-v") == 0 || strcmp(argv[1], "--version") == 0) {
            printf("UwULang %s\n", VERSION);
            return 0;
        }

        // if -h or --help
        if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0) {
            printf("UwULang %s\n", VERSION);
            printf("Usage: uwulang [file]\n");
            printf("Options:\n");
            printf("  -h, --help\t\tShow this help message\n");
            printf("  -v, --version\t\tShow version\n");
            return 0;
        }

        event_loop(argc, argv);
    }

    return 0;
}
